# Processing all human chromosomes
## 9.5.2024 (m.d.y)
## chrXXXX.fna.gz.ipynb

### Chromosome assemblies

1. [https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.40_GRCh38.p14/38p14_assembly_structure/Primary_Assembly/assembled_chromosomes/FASTA/](https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.40_GRCh38.p14/GCF_000001405.40_GRCh38.p14_assembly_structure/Primary_Assembly/assembled_chromosomes/FASTA/)

In [24]:
# GLOBALS
import sys
import json
import gzip
from hashlib import sha256
import os

inputPath = 'C:\\Users\\creeperpandatrex\\Documents\\1000genomes\\DATA\\GCF000001405-38p14-assembly\\'

inputChrom = [10,11,12,13,14,15,16,17,18,19,20,21,22,1,2,3,4,5,6,7,8,9,24] # chrom X is 23 chrom Y is 24
inputFileArr = [
"chr10.fna.gz",
"chr11.fna.gz",
"chr12.fna.gz",
"chr13.fna.gz",
"chr14.fna.gz",
"chr15.fna.gz",
"chr16.fna.gz",
"chr17.fna.gz",
"chr18.fna.gz",
"chr19.fna.gz",
"chr20.fna.gz",
"chr21.fna.gz",
"chr22.fna.gz",
"chr1.fna.gz",
"chr2.fna.gz",
"chr3.fna.gz",
"chr4.fna.gz",
"chr5.fna.gz",
"chr6.fna.gz",
"chr7.fna.gz",
"chr8.fna.gz",
"chr9.fna.gz",
"chrX.fna.gz"
]
def idSymbolBuildSet(myPath):
    print(myPath)
    myList = []
    obj = open(myPath, 'r')
    for i in obj:
        local = i.split()
        myList.append(local)
        local = []
    print(len(myList)) # 1523
    mySet = set(tuple(i) for i in myList) # remove duplicates 1329
    set2list = list(mySet)
    return set2list

def processChrWithIdSymbolSanityCheck():
    myListOfDicMap = []
    myListIdSymbSeqHash = []
    sanityCheck = []
    primer = ''
    for i in range(len(tmp)):
        localArray = []
        mySeed = os.getenv('SEED')
        myDicMap = {}
        myDicTupleSeqHash = {}
        localTuple = tmp[i]
        localTupleHash = sha256((convertTuple(localTuple) + mySeed).encode()).hexdigest()
        myDicMap.update({'fourTuple':localTuple, 'fourTupleHash':localTupleHash})
        myListOfDicMap.append(myDicMap)
        seqHash = buildSeqHash(localTuple, mySeed) # [myseqhash, myseq]
        myDicTupleSeqHash.update( {localTupleHash: seqHash[0] })
        # print(type(seqHash[1]))
        seqSize = len(seqHash[1])
        # print('>'+str(localTuple)+'\t'+str(seqSize)+'\n')
        primer += '>'+str(localTuple)+'\t'+str(seqSize)+'\n'
        first = ''
        last = ''
        counter = 0
        if len(seqHash[1]) > 40:
            for j in range(40):
                first += seqHash[1][counter]
                counter += 1
            localArray.append(first)
        counter =  -40
        if len(seqHash[1]) > 40:
            for k in range(40):
                last += (seqHash[1][counter])
                counter += 1
            localArray.append(last)
        # print(first+'..'+len(seqHash)+'..'+last)   #### PRINT
        # print(first+'....'+last)   #########  PRINT
        primer += first+'....'+last+'\n'
        sanityCheck.append(seqHash[1])
        myListIdSymbSeqHash.append(myDicTupleSeqHash)
    return [myListOfDicMap, myListIdSymbSeqHash, sanityCheck, primer]    
    
def print2file (myPath, result):
    out = open(myPath, 'w')
    out.write(result) # it must be a string

def buildSeqHash(tuple, mySeed):
    mySeq = ''
    id,symb,start,end = tuple
    start = int(start) - 1
    end = int(end) 
    # counter = start + 80
    # counter = start + inputIndex[CURRENT_CHROMO]
    counter = start
    for i in range(end - start):
        mySeq += my[counter]
        counter += 1
    mySeqHash = sha256(((mySeq+mySeed)).encode()).hexdigest()
    return [mySeqHash, mySeq] # both string


# print(buildSeq(processing[0]['fourTuple']))
        
def convertTuple(tup):
    str = ''.join(tup)
    return str

def print2gz (myPath, result, addExt):
   with gzip.open(myPath + addExt, 'w') as f:
    f.write((json.dumps(result)).encode()) # no decode() since it is a str already


In [26]:
# my is a string of chromosome 'NN' fasta files with 80 char per line to fasta with 2 lines ONLY

def fasta2one (myPath):
    with myPath as file:
        build = ''
        counter = 0
        print(myPath)
        for i in file:
            if counter != 0:
                build += i.strip()
            counter = 1
        return build
index = 0
for item in inputFileArr:
    print(item)
    my = fasta2one(gzip.open(inputPath + item, 'rt')) # 9.4.2024 (m.d.y) in 58 seconds
    print('len(my): ', len(my))
    myLocal = inputPath + 'awk_ID_symbol_start_end_chr_' + str(inputChrom[index])
    print('myLocal: ', myLocal )
    tmp = idSymbolBuildSet(myLocal)
    processing = processChrWithIdSymbolSanityCheck()
    result = processing
    print2gz(inputPath + item , result[0], '.labelmap.v3.gz' )
    print2gz(inputPath + item , result[1], '.labelseqhash.v3.gz' )
    print2file(inputPath + item + '.v3.primer', processing[3])
    print(len(processing[2]))
    index += 1



chr10.fna.gz
<_io.TextIOWrapper name='C:\\Users\\creeperpandatrex\\Documents\\1000genomes\\DATA\\GCF000001405-38p14-assembly\\chr10.fna.gz' encoding='utf-8'>
len(my):  133797422
myLocal:  C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_10
C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_10
2726
2427
chr11.fna.gz
<_io.TextIOWrapper name='C:\\Users\\creeperpandatrex\\Documents\\1000genomes\\DATA\\GCF000001405-38p14-assembly\\chr11.fna.gz' encoding='utf-8'>
len(my):  135086622
myLocal:  C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_11
C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_11
3468
3235
chr12.fna.gz
<_io.TextIOWrapper name='C:\\Users\\creeperpandatrex\\Documents\\1000genomes\\DATA\\GCF000001405-38p14-assembly\\chr12.fna.gz' encoding='utf-8'

# Output
1. ## chr10.fna.gz
1. <_io.TextIOWrapper name='C:\\Users\\creeperpandatrex\\Documents\\1000genomes\\DATA\\GCF000001405-38p14-assembly\\chr10.fna.gz' encoding='utf-8'>
1. len(my):  133797422
1. myLocal:  C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_10
1. C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_10
1. 2726
1. 2427
1. ## chr11.fna.gz
1. <_io.TextIOWrapper name='C:\\Users\\creeperpandatrex\\Documents\\1000genomes\\DATA\\GCF000001405-38p14-assembly\\chr11.fna.gz' encoding='utf-8'>
1. len(my):  135086622
1. myLocal:  C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_11
1. C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_11
1. 3468
1. 3235
1. ## chr12.fna.gz
1. <_io.TextIOWrapper name='C:\\Users\\creeperpandatrex\\Documents\\1000genomes\\DATA\\GCF000001405-38p14-assembly\\chr12.fna.gz' encoding='utf-8'>
1. len(my):  133275309
1. myLocal:  C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_12
1. C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_12
1. 3070
1. 2853
1. ## chr13.fna.gz
1. <_io.TextIOWrapper name='C:\\Users\\creeperpandatrex\\Documents\\1000genomes\\DATA\\GCF000001405-38p14-assembly\\chr13.fna.gz' encoding='utf-8'>
1. len(my):  114364328
1. myLocal:  C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_13
1. C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_13
1. 1591
1. 1542
1. ## chr14.fna.gz
1. <_io.TextIOWrapper name='C:\\Users\\creeperpandatrex\\Documents\\1000genomes\\DATA\\GCF000001405-38p14-assembly\\chr14.fna.gz' encoding='utf-8'>
1. len(my):  107043718
1. myLocal:  C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_14
1. C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_14
1. 2372
1. 2232
1. ## chr15.fna.gz
1. <_io.TextIOWrapper name='C:\\Users\\creeperpandatrex\\Documents\\1000genomes\\DATA\\GCF000001405-38p14-assembly\\chr15.fna.gz' encoding='utf-8'>
1. len(my):  101991189
1. myLocal:  C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_15
1. C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_15
1. 2257
1. 2040
1. ## chr16.fna.gz
1. <_io.TextIOWrapper name='C:\\Users\\creeperpandatrex\\Documents\\1000genomes\\DATA\\GCF000001405-38p14-assembly\\chr16.fna.gz' encoding='utf-8'>
1. len(my):  90338345
1. myLocal:  C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_16
1. C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_16
1. 2408
1. 2231
1. ## chr17.fna.gz
1. <_io.TextIOWrapper name='C:\\Users\\creeperpandatrex\\Documents\\1000genomes\\DATA\\GCF000001405-38p14-assembly\\chr17.fna.gz' encoding='utf-8'>
1. len(my):  83257441
1. myLocal:  C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_17
1. C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_17
1. 2987
1. 2728
1. ## chr18.fna.gz
1. <_io.TextIOWrapper name='C:\\Users\\creeperpandatrex\\Documents\\1000genomes\\DATA\\GCF000001405-38p14-assembly\\chr18.fna.gz' encoding='utf-8'>
1. len(my):  80373285
1. myLocal:  C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_18
1. C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_18
1. 1156
1. 1118
1. ## chr19.fna.gz
1. <_io.TextIOWrapper name='C:\\Users\\creeperpandatrex\\Documents\\1000genomes\\DATA\\GCF000001405-38p14-assembly\\chr19.fna.gz' encoding='utf-8'>
1. len(my):  58617616
1. myLocal:  C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_19
1. C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_19
1. 2921
1. 2701
1. ## chr20.fna.gz
1. <_io.TextIOWrapper name='C:\\Users\\creeperpandatrex\\Documents\\1000genomes\\DATA\\GCF000001405-38p14-assembly\\chr20.fna.gz' encoding='utf-8'>
1. len(my):  64444167
1. myLocal:  C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_20
1. C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_20
1. 1661
1. 1488
1. ## chr21.fna.gz
1. <_io.TextIOWrapper name='C:\\Users\\creeperpandatrex\\Documents\\1000genomes\\DATA\\GCF000001405-38p14-assembly\\chr21.fna.gz' encoding='utf-8'>
1. len(my):  46709983
1. myLocal:  C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_21
1. C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_21
1. 896
1. 854
1. ## chr22.fna.gz
1. <_io.TextIOWrapper name='C:\\Users\\creeperpandatrex\\Documents\\1000genomes\\DATA\\GCF000001405-38p14-assembly\\chr22.fna.gz' encoding='utf-8'>
1. len(my):  50818468
1. myLocal:  C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_22
1. C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_22
1. 1523
1. 1329
1. ## chr1.fna.gz
1. <_io.TextIOWrapper name='C:\\Users\\creeperpandatrex\\Documents\\1000genomes\\DATA\\GCF000001405-38p14-assembly\\chr1.fna.gz' encoding='utf-8'>
1. len(my):  248956422
1. myLocal:  C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_1
1. C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_1
1. 31108
1. 28828
1. ## chr2.fna.gz
1. <_io.TextIOWrapper name='C:\\Users\\creeperpandatrex\\Documents\\1000genomes\\DATA\\GCF000001405-38p14-assembly\\chr2.fna.gz' encoding='utf-8'>
1. len(my):  242193529
1. myLocal:  C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_2
1. C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_2
1. 8745
1. 7997
1. ## chr3.fna.gz
1. <_io.TextIOWrapper name='C:\\Users\\creeperpandatrex\\Documents\\1000genomes\\DATA\\GCF000001405-38p14-assembly\\chr3.fna.gz' encoding='utf-8'>
1. len(my):  198295559
1. myLocal:  C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_3
1. C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_3
1. 3628
1. 3393
1. ## chr4.fna.gz
1. <_io.TextIOWrapper name='C:\\Users\\creeperpandatrex\\Documents\\1000genomes\\DATA\\GCF000001405-38p14-assembly\\chr4.fna.gz' encoding='utf-8'>
1. len(my):  190214555
1. myLocal:  C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_4
1. C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_4
1. 2879
1. 2760
1. ## chr5.fna.gz
1. <_io.TextIOWrapper name='C:\\Users\\creeperpandatrex\\Documents\\1000genomes\\DATA\\GCF000001405-38p14-assembly\\chr5.fna.gz' encoding='utf-8'>
1. len(my):  181538259
1. myLocal:  C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_5
1. C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_5
1. 3362
1. 2958
1. ## chr6.fna.gz
1. <_io.TextIOWrapper name='C:\\Users\\creeperpandatrex\\Documents\\1000genomes\\DATA\\GCF000001405-38p14-assembly\\chr6.fna.gz' encoding='utf-8'>
1. len(my):  170805979
1. myLocal:  C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_6
1. C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_6
1. 3699
1. 3451
1. ## chr7.fna.gz
1. <_io.TextIOWrapper name='C:\\Users\\creeperpandatrex\\Documents\\1000genomes\\DATA\\GCF000001405-38p14-assembly\\chr7.fna.gz' encoding='utf-8'>
1. len(my):  159345973
1. myLocal:  C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_7
1. C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_7
1. 3406
1. 3111
1. ## chr8.fna.gz
1. <_io.TextIOWrapper name='C:\\Users\\creeperpandatrex\\Documents\\1000genomes\\DATA\\GCF000001405-38p14-assembly\\chr8.fna.gz' encoding='utf-8'>
1. len(my):  145138636
1. myLocal:  C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_8
1. C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_8
1. 2596
1. 2444
1. ## chr9.fna.gz
1. <_io.TextIOWrapper name='C:\\Users\\creeperpandatrex\\Documents\\1000genomes\\DATA\\GCF000001405-38p14-assembly\\chr9.fna.gz' encoding='utf-8'>
1. len(my):  138394717
1. myLocal:  C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_9
1. C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_9
1. 2734
1. 2584
1. ## chrX.fna.gz
1. <_io.TextIOWrapper name='C:\\Users\\creeperpandatrex\\Documents\\1000genomes\\DATA\\GCF000001405-38p14-assembly\\chrX.fna.gz' encoding='utf-8'>
1. len(my):  156040895
1. myLocal:  C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_24
1. C:\Users\creeperpandatrex\Documents\1000genomes\DATA\GCF000001405-38p14-assembly\awk_ID_symbol_start_end_chr_24
1. 2512
1. 2375
